[inline-modules][docs] Add reference and tutorial for inline modules (#44110)
expo/expo: docs/pages/modules/inline-modules-tutorial.mdx

---
title: 'Tutorial: Create an inline module'
sidebar_title: Create an inline module
description: A tutorial on creating inline modules.
---

import { Grid01Icon } from '@expo/styleguide-icons/outline/Grid01Icon';

import { BoxLink } from '~/ui/components/BoxLink';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';

> In this tutorial, you'll build a simple inline module and an inline view.

Reviewer: ```suggestion
In this tutorial, you create a native module and a native view directly inside your Expo project directory using inline modules. Unlike the standard Expo Modules, inline modules don't require a separate package or `create-expo-module` scaffolding. You write Kotlin and Swift files alongside your app code, and Expo discovers them automatically.
```

This doesn't define why one should use this or what it does. Since we are adding this as a new concept, what a developer will get from it should be defined here in the introduction.

Suggestion to expand this.

<Step label="1">

## Setup your project

Open the `app.json` and set the `expo.experiments.inlineModules.watchedDirectories` to `["app"]`

```json app.json
{
  "expo": {
    "experiments": {
      "inlineModules": {
        "watchedDirectories": ["app"]
      }
    }
  }
}
```

Defining `expo.experiments.inlineModules` enables inline modules functionalities in Expo. Note that you will need to prebuild the ios after changing this option as it changes Xcode project files!

In `expo.experiments.inlineModules.watchedDirectories` you can specify in which directories your inline modules lie. Note that not all directories are allowed, checkout the [inline modules reference](/modules/inline-modules-reference) for more information.

</Step>

<Step label="2">

## Run prebuild

Run the prebuild command to generate android and ios projects with the inline modules setup.

<Terminal cmdCopy="yarn expo prebuild" cmd={['$ yarn expo prebuild']} />

</Step>

<Step label="3">

## Create an inline module

> Create a kotlin file `FirstInlineModule.kt` in the `app` directory and write a module in there.

Reviewer: ```suggestion
For Android, create **FirstInlineModule.kt** inside the **app** directory and write a module similar to the following:
```

We use bold text format for directory and file names as per our wrting styleguide: https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#file-names-directory-names-file-extensions-as-bold-text

I am using the **app** directory here, which is where the app's navigation is defined when using Expo Router.  Let me know if my assumption is wrong.

```kotlin app/FirstInlineModule.kt
package app

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class FirstInlineModule : Module() {
  override fun definition() = ModuleDefinition {
    Constant("Hello") { ->
      "Hello Android inline modules!"
    }
  }
}

```

Also create a `FirstInlineModule.swift` in the `app` directory and write a similar module.

```swift app/FirstInlineModule.swift
import ExpoModulesCore

public class FirstInlineModule: Module {
  public func definition() -> ModuleDefinition {
    Constant("Hello") {
      return "Hello Ios inline modules!"
    }
  }
}

```

</Step>

<Step label="4">

## Use the module in JS

In TS/JS you can use the module in a following way

```tsx app/index.tsx
import { requireNativeModule } from 'expo';
import { Text } from 'react-native';

const FirstInlineModule = requireNativeModule('FirstInlineModule');

export default function InlineModulesDemoComponent() {
  return <Text> {FirstInlineModule.Hello} </Text>;
}
```

You can run you android app

<Terminal cmdCopy="yarn expo run:android" cmd={['$ yarn expo run:android']} />

Or your ios app

> <Terminal cmdCopy="yarn expo run:ios" cmd={['$ yarn expo run:ios']} />

Reviewer: ```suggestion
Now, you can run the example app:

<Terminal
  cmd={[
    '# Run the example app on Android',
    '$ npx expo run:android',
    '',
    '# Run the example app on iOS',
    '$ npx expo run:ios',
  ]}
/>
```

        }
      }
    }
  }
}

class ExpoWebView: ExpoView, WKNavigationDelegate {
  let webView = WKWebView()
  let onLoad = EventDispatcher()

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = true
    webView.navigationDelegate = self
    addSubview(webView)
  }

  override func layoutSubviews() {
    webView.frame = bounds
  }

  func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
    if let url = webView.url {
      onLoad([
        "url": url.absoluteString
      ])
    }
  }
}

```

</Step>

<Step label="6">
## Use the native view in JS
You use the inline view in a similar way to the inline module:

```tsx app/index.tsx
import { requireNativeModule, requireNativeView } from 'expo';
import { Text } from 'react-native';

const FirstInlineModule = requireNativeModule('FirstInlineModule');
const FirstInlineView = requireNativeView('FirstInlineView');

export default function InlineModulesDemoComponent() {
  return (
    <>
      <Text> {FirstInlineModule.Hello} </Text>
      <FirstInlineView style={{ flex: 1 }} url="https://docs.expo.dev/modules/" />
    </>
  );
}
```

> 

Author: That's right, changed it a bit.
